boot/assets.go

```diff
@@ -20,27 +20,128 @@
 package boot
 
 import (
+	"crypto"
+	_ "crypto/sha256"
+	"encoding/hex"
+	"fmt"
+	"io"
+	"os"
+	"path/filepath"
+
 	"github.com/snapcore/snapd/asserts"
+	"github.com/snapcore/snapd/bootloader"
+	"github.com/snapcore/snapd/dirs"
 	"github.com/snapcore/snapd/gadget"
+	"github.com/snapcore/snapd/osutil"
+	"github.com/snapcore/snapd/strutil"
 )
 
+type trustedAssetsCache struct {
+	cacheDir string
+	hash     crypto.Hash
+}
+
+func newTrustedAssetsCache(cacheDir string) *trustedAssetsCache {
+	return &trustedAssetsCache{cacheDir: cacheDir, hash: crypto.SHA256}
+}
+
+func (c *trustedAssetsCache) assetKey(blName, assetName, assetHash string) string {
+	return filepath.Join(blName, fmt.Sprintf("%s-%s", assetName, assetHash))
+}
+
+func (c *trustedAssetsCache) tempAssetKey(blName, assetName string) string {
+	return filepath.Join(blName, assetName+".temp")
+}
+
+func (c *trustedAssetsCache) pathInCache(part string) string {
+	return filepath.Join(c.cacheDir, part)
+}
+
+// Add entry for
+func (c *trustedAssetsCache) Add(assetPath, blName, assetName string) (ta *trackedAsset, added bool, err error) {
+	if err := os.MkdirAll(c.pathInCache(blName), 0755); err != nil {
+		return nil, false, fmt.Errorf("cannot create cache directory: %v", err)
+	}
+
+	// input
+	inf, err := os.Open(assetPath)
+	if err != nil {
+		return nil, false, fmt.Errorf("cannot open asset file: %v", err)
+	}
+	// temporary output
+	tempPath := c.pathInCache(c.tempAssetKey(blName, assetName))
+	outf, err := osutil.NewAtomicFile(tempPath, 0644, 0, osutil.NoChown, osutil.NoChown)
+	if err != nil {
+		return nil, false, fmt.Errorf("cannot create temporary cache file: %v", err)
+	}
+	defer outf.Cancel()
+
+	// copy and hash at the same time
+	h := c.hash.New()
+	tr := io.TeeReader(inf, h)
+	if _, err := io.Copy(outf, tr); err != nil {
+		return nil, false, fmt.Errorf("cannot copy trusted asset to cache: %v", err)
+	}
+	hashStr := hex.EncodeToString(h.Sum(nil))
+	cacheKey := c.assetKey(blName, assetName, hashStr)
+
+	ta = &trackedAsset{
+		blName: blName,
+		name:   assetName,
+		hash:   hashStr,
+	}
+
+	targetName := c.pathInCache(cacheKey)
+	if osutil.FileExists(targetName) {
+		// asset is already cached
+		return ta, false, nil
+	}
+	// commit under a new name
+	if err := outf.CommitAs(targetName); err != nil {
+		return nil, false, fmt.Errorf("cannot commit file to assets cache: %v", err)
+	}
+	return ta, true, nil
+}
+
+func (c *trustedAssetsCache) Drop(blName, assetName, hashStr string) error {
```

Comment:
this should be called Remove (we mostly use drop in the context of dropping privileges)

Reply:
/me silently thinks about Evict() since it's a cache, but Remove works too.